packages/@glimmer/reference/lib/template.ts

```diff
@@ -307,8 +307,12 @@ export class IterationItemReference<T = unknown> implements TemplatePathReferenc
   }
 
   update(value: T) {
-    dirtyTag(this.tag);
-    this.itemValue = value;
+    let { itemValue } = this;
+
+    if (value !== itemValue) {
```

Comment:
Can you remind me what `itemValue` actually represents? Is it the value that we find after looking up the `key` path (or `guidFor` if `key="@identity"`)? Or is it literally the item being yielded?

Reply:
> Or is it literally the item being yielded?

This. If the "iterable" is an array, then for each value in the array you have an `IterationItemReference` where `itemValue` is the value.

Reply:
Shouldn’t it be the key that we are checking for stability (not the item itself)? I realize that in the default case for Ember there is no difference between the two (since Ember uses `key=“@identity”` by default), but it is possible to specify a key and AFAICT that is the thing that is important to check for stability :thinking:.

For example, imagine the case where the array and its items are always regenerated (e.g. the immutable pattern) so you’ve used `key` to tell `{{#each` that it should care about some stable property on the item (e.g. `id`).

Reply:
AFAICT we're not checking the stability here. The key is used when [checking the stability of the array itself over here]() – checking which items were added, which were moved, and which stayed in the same place. If the array is touched and an item is considered retained (by comparing keys), then it just always updates the value and the memo (in the case of arrays, the array index). If you don't compare the values at this point, doing anything to the array can become quite an expensive operation O(n). 

There's more though.

If you have an array of objects and modify some property deep inside, then all of this iterator code is skipped and just that nested `PropertyReference` is updated. So the key you provide to `{{each}}` plays no role in that case. But doing anything to array will trigger the iterator synchronization/list revalidation code and call this update method. So we do all this extra work, even though in some cases Ember provides the exact key modified via `dirtyTagFor`.

Also, `{{each-in}}` is possibly even more broken. Since the default key is `@identity`, updating a the value of an existing key won't cause an update, as may be expected by an end user. It will actually destroy and recreate the item. So, again, if you're using render modifier, you'll get a bunch of weird triggers.

TBH I don't understand why we should be comparing values here at all. Aren't there tags for every value anyways? Shouldn't we be checking those instead?

Reply:
Gotcha, glad you dug in here :smile:.

@pzuraq - Any reason we are checking `===` vs a tag validation?

Reply:
It's the opposite in this case - we know that the value has possibly changed, and we need to dirty the tag if so and update it. We don't have another tag to check.